plugins/chat/spec/services/chat/remove_user_from_channel_spec.rb

```diff
@@ -0,0 +1,121 @@
+# frozen_string_literal: true
+
+RSpec.describe Chat::RemoveUserFromChannel do
+  describe described_class::Contract, type: :model do
+    subject(:contract) { described_class.new(user_id: nil, channel_id: nil) }
+
+    it { is_expected.to validate_presence_of :channel_id }
+    it { is_expected.to validate_presence_of :user_id }
+  end
+
+  describe ".call" do
+    subject(:result) { described_class.call(params:, **dependencies) }
+
+    fab!(:acting_user) { Fabricate(:admin) }
+    fab!(:user)
+    fab!(:channel) { Fabricate(:chat_channel) }
+
+    let(:guardian) { Guardian.new(acting_user) }
+    let(:params) { { channel_id: channel.id, user_id: user.id } }
+    let(:dependencies) { { guardian: } }
+
+    context "when all steps pass" do
+      context "when category channel" do
+        context "with existing membership" do
+          before do
+            channel.add(user)
+            channel.add(acting_user)
+            Chat::Channel.ensure_consistency!
+          end
+
+          it { is_expected.to run_successfully }
+
+          it "unfollows the channel" do
+            membership = channel.membership_for(user)
+
+            expect { result }.to change { membership.reload.following }.from(true).to(false)
+          end
+
+          it "recomputes user count" do
+            expect { result }.to change { channel.reload.user_count }.from(2).to(1)
+          end
+        end
+
+        context "with no existing membership" do
+          it { is_expected.to run_successfully }
+
+          it "does nothing" do
+            expect { result }.to_not change { Chat::UserChatChannelMembership }
+          end
+        end
+      end
+
+      context "when group channel" do
+        context "with existing membership" do
+          fab!(:channel) do
+            Fabricate(:direct_message_channel, group: true, users: [acting_user, user])
+          end
+
+          before { Chat::Channel.ensure_consistency! }
+
+          it { is_expected.to run_successfully }
+
+          it "leaves the channel" do
+            membership = channel.membership_for(user)
+
+            result
+
+            expect(Chat::UserChatChannelMembership.exists?(membership.id)).to eq(false)
+            expect(channel.chatable.direct_message_users.where(user_id: user.id).exists?).to eq(
+              false,
+            )
+          end
+
+          it "recomputes user count" do
+            expect { result }.to change { channel.reload.user_count }.from(2).to(1)
+          end
+        end
+
+        context "with no existing membership" do
+          it { is_expected.to run_successfully }
+
+          it "does nothing" do
+            expect { result }.to_not change { Chat::UserChatChannelMembership }
+          end
+        end
+      end
+
+      context "when direct channel" do
```

Comment:
Maybe we could make this a bit more explicit by saying something like `when one-on-one DM`?